gossipsub.go

```diff
@@ -216,10 +218,88 @@ func NewGossipSubWithRouter(ctx context.Context, h host.Host, rt PubSubRouter, o
 	return NewPubSub(ctx, h, rt, opts...)
 }
 
+// ReadOnlyRoutingMetadata is a read-only view of the routing metadataTracer maintained by the GossipSubRouter.
+// It is used to expose the metadataTracer to the application without exposing the router internals.
+type ReadOnlyRoutingMetadata struct {
+	mu sync.RWMutex
+	// metda-date will be updated every k-heartbeats by the router. This is used to avoid unnecessary updates.
+	k uint64
+	// heartbeat counter
+	count  uint64
+	mesh   map[string]map[peer.ID]struct{}
+	fanout map[string]map[peer.ID]struct{}
+}
+
+func newReadOnlyRoutingMetadataTracer(k uint64) *ReadOnlyRoutingMetadata {
+	return &ReadOnlyRoutingMetadata{
+		k:      k,
+		count:  0,
+		mesh:   make(map[string]map[peer.ID]struct{}),
+		fanout: make(map[string]map[peer.ID]struct{}),
+	}
+}
+
+func (r *ReadOnlyRoutingMetadata) getMeshPeers(topic string) []peer.ID {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	peers := make([]peer.ID, 0, len(r.mesh[topic]))
+	for p := range r.mesh[topic] {
+		peers = append(peers, p)
+	}
+	return peers
+}
+
+func (r *ReadOnlyRoutingMetadata) getFanoutPeers(topic string) []peer.ID {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	peers := make([]peer.ID, 0, len(r.fanout[topic]))
+	for p := range r.fanout[topic] {
+		peers = append(peers, p)
+	}
+	return peers
+}
+
+func (r *ReadOnlyRoutingMetadata) update(mesh map[string]map[peer.ID]struct{}, fanout map[string]map[peer.ID]struct{}) {
+	r.count++
+	if r.count%r.k != 0 {
+		return // skip update
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// deep copies mesh and fanout maps
+	for topic, peers := range mesh {
+		r.mesh[topic] = make(map[peer.ID]struct{}, len(peers))
+		for p := range peers {
+			r.mesh[topic][p] = struct{}{}
+		}
+	}
+
+	for topic, peers := range fanout {
+		r.fanout[topic] = make(map[peer.ID]struct{}, len(peers))
+		for p := range peers {
+			r.fanout[topic][p] = struct{}{}
+		}
+	}
+}
+
+type GossipSubRouterOption func(*GossipSubRouter)
```

Comment:
the general pattern is to make a pubsub option that checks the router and asserts it to be gs.
Lets keep doing it that way.